scan/base.go

```diff
@@ -40,6 +44,358 @@ type base struct {
+			return
+		}
+		if jsonError.Error == "HTTP Token: Access denied.\n" {
+			return vinfos, fmt.Errorf("wordpress: HTTP Token: Access denied")
+		}
+		return vinfos, fmt.Errorf("status: %s", resp.Status)
+	}
+	coreConvertVinfos(string(body))
+	return
+}
+
+func coreConvertVinfos(stdout string) (vinfos []models.VulnInfo, err error) {
+	data := map[string]WpCveInfos{}
+	if err = json.Unmarshal([]byte(stdout), &data); err != nil {
+		var jsonError WpCveInfos
+		if err = json.Unmarshal([]byte(stdout), &jsonError); err != nil {
+			return
+		}
+	}
+	for _, e := range data {
+		if len(e.Vulnerabilities) == 0 {
+			continue
+		}
+		for _, vulnerability := range e.Vulnerabilities {
+			if len(vulnerability.References.Cve) == 0 {
+				continue
+			}
+			notFixedYet := false
+			if len(vulnerability.FixedIn) == 0 {
+				notFixedYet = true
+			}
+			var cveIDs []string
+			for _, cveNumber := range vulnerability.References.Cve {
+				cveIDs = append(cveIDs, "CVE-"+cveNumber)
+			}
+
+			for _, cveID := range cveIDs {
+				vinfos = append(vinfos, models.VulnInfo{
+					CveID: cveID,
+					CveContents: models.NewCveContents(
+						models.CveContent{
+							CveID: cveID,
+							Title: vulnerability.Title,
+						},
+					),
+					AffectedPackages: models.PackageStatuses{
+						{
+							NotFixedYet: notFixedYet,
+						},
+					},
+				})
+			}
+		}
+	}
+	return
+}
+
+//WpStatus is for wp command
+type WpStatus struct {
+	Name    string `json:"name"`
+	Status  string `json:"status"`
+	Update  string `json:"update"`
+	Version string `json:"version"`
+}
+
+func detectWpTheme(c *base) (vinfos []models.VulnInfo, err error) {
+	cmd := fmt.Sprintf("wp theme list --path=%s --format=json", c.ServerInfo.WpPath)
+
+	var themes []WpStatus
+	var r execResult
+	if r = exec(c.ServerInfo, cmd, noSudo); r.isSuccess() {
+		if err = json.Unmarshal([]byte(r.Stdout), &themes); err != nil {
+			return
+		}
+	}
+	if !r.isSuccess() {
+		return vinfos, fmt.Errorf("%s", cmd)
+	}
+
+	for _, theme := range themes {
+		url := fmt.Sprintf("https://wpvulndb.com/api/v3/themes/%s", theme.Name)
+		contentHttpRequest(c, theme, url)
+	}
+	return
+}
+
+func detectWpPlugin(c *base) (vinfos []models.VulnInfo, err error) {
+	cmd := fmt.Sprintf("wp plugin list --path=%s --format=json", c.ServerInfo.WpPath)
+
+	var plugins []WpStatus
+	var r execResult
+	if r := exec(c.ServerInfo, cmd, noSudo); r.isSuccess() {
+		if err = json.Unmarshal([]byte(r.Stdout), &plugins); err != nil {
+			return
+		}
+	}
+	if !r.isSuccess() {
+		return vinfos, fmt.Errorf("%s", cmd)
+	}
+
+	for _, plugin := range plugins {
+		url := fmt.Sprintf("https://wpvulndb.com/api/v3/plugins/%s", plugin.Name)
+		var tmpVinfos []models.VulnInfo
+		if tmpVinfos, err = contentHttpRequest(c, plugin, url); err != nil {
+			return
+		}
+		vinfos = append(vinfos, tmpVinfos...)
+	}
+	return
+}
+
+func contentHttpRequest(c *base, content WpStatus, url string) (vinfos []models.VulnInfo, err error) {
```

Comment:
__[Golint]__ func contentHttpRequest should be contentHTTPRequest

<sub>posted by Sider</sub>